src/driver-roccat-kone-emp.c

```diff
@@ -0,0 +1,1132 @@
+/* FIXME:   { 171, Control panel }, */
+/* FIXME:   { 172, System settings }, */
+/* FIXME:   { 173, Task Manager }, */
+/* FIXME:   { 174, Screen settings }, */
+/* FIXME:   { 175, Screensaver settings }, */
+/* FIXME:   { 176, Themes }, */
+/* FIXME:   { 177, Date and Time }, */
+/* FIXME:   { 178, Network settings }, */
+/* FIXME:   { 179, Admin settings }, */
+/* FIXME:   { 180, Firewall }, */
+/* FIXME:   { 181, Regedit }, */
+/* FIXME:   { 182, Event monitor }, */
+/* FIXME:   { 183, Performance monitor }, */
+/* FIXME:   { 184, Audio settings }, */
+/* FIXME:   { 185, Internet settings }, */
+/* FIXME:   { 186, Directx diagnostics }, */
+/* FIXME:   { 187, Command line }, */
+/* FIXME:   { 188, System poweroff }, */
+/* FIXME:   { 189, System sleep }, */
+/* FIXME:   { 190, System wakeup }, */
+
+/* FIXME:   { 191, Set profile 1 }, */
+/* FIXME:   { 192, Set profile 2 }, */
+/* FIXME:   { 193, Set profile 3 }, */
+/* FIXME:   { 194, Set profile 4 }, */
+/* FIXME:   { 195, Set profile 5 }, */
+};
+
+static const struct ratbag_button_action*
+roccat_raw_to_button_action(uint8_t data)
+{
+	struct roccat_button_mapping *mapping;
+
+	ARRAY_FOR_EACH(roccat_button_mapping, mapping) {
+		if (mapping->raw == data)
+			return &mapping->action;
+	}
+
+	return NULL;
+}
+
+static uint8_t
+roccat_button_action_to_raw(const struct ratbag_button_action *action)
+{
+	struct roccat_button_mapping *mapping;
+
+	ARRAY_FOR_EACH(roccat_button_mapping, mapping) {
+		if (ratbag_button_action_match(&mapping->action, action))
+			return mapping->raw;
+	}
+
+	return 0;
+}
+
+static inline uint16_t
+roccat_get_unaligned_u16(uint8_t *buf)
+{
+	return (buf[1] << 8) | buf[0];
+}
+
+/**
+ * Compute the CRC from buf
+ * len should be the length of buf, including the two bytes used for CRC
+ */
+static inline uint16_t
+roccat_compute_crc(uint8_t *buf, unsigned int len)
+{
+	unsigned i;
+	uint16_t crc = 0;
+
+	if (len < 3)
+		return 0;
+
+	for (i = 0; i < len - 2; i++) {
+		crc += buf[i];
+	}
+
+	return crc;
+}
+
+/**
+ * Returns if the CRC in buf is valid.
+ * The CRC is expected to be the last two bytes of buf
+ * len should be the length of buf, including the CRC
+ */
+static inline int
+roccat_crc_is_valid(struct ratbag_device *device, uint8_t *buf, unsigned int len)
+{
+	uint16_t crc;
+	uint16_t given_crc;
+
+	if (len < 3)
+		return 0;
+
+	crc = roccat_compute_crc(buf, len);
+
+	given_crc = roccat_get_unaligned_u16(&buf[len - 2]);
+
+	log_raw(device->ratbag,
+		"checksum computed: 0x%04x, checksum given: 0x%04x - %s\n",
+		crc,
+		given_crc,
+		crc == given_crc ? "OK" : "FAIL");
+
+
+	return crc == given_crc;
+}
+
+static int
+roccat_is_ready(struct ratbag_device *device)
+{
+	uint8_t buf[3] = { 0 };
+	int rc;
+
+	rc = ratbag_hidraw_get_feature_report(device, ROCCAT_REPORT_ID_CONFIGURE_PROFILE,
+						  buf, sizeof(buf));
+	if (rc < 0)
+		return rc;
+	if (rc != sizeof(buf))
+		return -EIO;
+
+	if (buf[1] == 0x03)
+		msleep(100);
+
+	if (buf[1] == 0x02)
+		return 2;
```

Comment:
What does 2 mean here? If it's an error type, you should probably add a `#define` and use it here and the function below.

Reply:
On this one, I don't have a clue. The functions `roccat_is_ready()` and `roccat_wait_ready()` are the one from _driver-roccat.c_